chainer/dataset/convert.py

```diff
@@ -211,6 +211,15 @@ def concat_examples(batch, device=None, padding=None):
     if not batch:
         raise ValueError('batch is empty')
 
+    # batch from TabularDataset.fetch
+    if isinstance(batch, tuple):
+        return tuple(
+            to_device(device, _concat_arrays(array, None)) for array in batch)
```

Comment:
It seems we can bypass `_concat_arrays` because `array` is just an `ndarray` (tabular array) in this place.

Reply:
`array` is not always an ndarray. It can be a list.

Reply:
I see. But `_concat_arrays` is originally for converting list of arrays in one column to an array, and causes another 5% overhead. The case you pointed out should be handled in a separate logic I think.

Reply:
Here, I want to convert a list of scalars/ndarrays into an ndarray. I think this behaviour is same as `_concat_arrays`. What is the intention of separated logic?

Reply:
Hmm, it seems there is three cases to handle input:

1. An ndarray (of scalars). It doesn’t need any concatanation. This is the typical case in which `TabularDataset` removes the performance overhead.

2. A list of scalars. It needs a conversion from a list to an ndarray. We just need the following lines in `_concat_arrays` (IMO it should be done in `fetch`):

```python
    if not isinstance(arrays[0], chainer.get_array_types()):
        arrays = numpy.asarray(arrays)
```

3. A list of N-dimension ndarrays. We should convert it to (N+1)-dimension ndarray done in `_concat_arrays` before. But it also can be done in `fetch` (if we update it’s signature as `None -> ndarray`).

Any ideas?

Reply:
Note that we have a chance to simplify the implementation of 3rd case if cupy/cupy#2081 is merged.

Reply:
In the term of performance,
even if we call concat in `fetch`, we cannot remove the overhead. 
Since the output of `get_examples` can be both list and ndarray. we should handle both cases. So the logic won't change.

In the term of design.
I think users can call `fetch` outside of updater. For example, users may want to check intermediate values for debugging. In that case, we cannot assume that all values can be concatenated.

Reply:
In addition, evaluation of some models are not require concatenated data and can accept list of arrays as its input. In that case, forcing concat in `fetch` is not good.

Reply:
I see. In this sense, does it make sense for you?

```suggestion
            to_device(device, (
                array if isinstance(array, chainer.get_array_types())
                else _concat_arrays(array, None)
            )) for array in batch)
```

Reply:
Ah, I see. I understand why you don't want to use `_concat_arrays` as is. I'll make another helper for this situation.

Reply:
I added `_as_array` that can bypass `ndarray` as is.